packages/react-reconciler/src/__tests__/ReactMemo-test.internal.js

```diff
@@ -177,6 +177,56 @@ describe('memo', () => {
         expect(ReactNoop.getChildren()).toEqual([span('Count: 1')]);
       });
 
+      it('bails out on props equality during context change', async () => {
+        const CountContext = React.createContext(0);
+
+        function readContext(Context) {
+          const dispatcher =
+            React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
+              .ReactCurrentDispatcher.current;
+          return dispatcher.readContext(Context);
+        }
+
+        function Inner(props) {
+          const renderCountRef = React.useRef(0);
+          readContext(CountContext);
+          return React.useMemo(() => {
+            const text = `Inner render count: ${++renderCountRef.current}`;
+            return <Text text={text} />;
+          }, [
+            props,
```

Comment:
Is this the only observable difference? Can you rewrite this test to check render count rather than whether `props` are referentially equal? Why does this matter at all?

Reply:
This checks the render count - see `renderCountRef`. I can't just count renders directly inside Inner, because it has to render (it reads changed context).

**Why this matters?**
Because it's inconsistent with "regular" memo component. Conceptually from the user's perspective `React.memo(Component)` & `React.memo(Component, shallowEqual)` should behave exactly the same (but they dont).

It makes caching assumptions less predictable. While this test is ofc artificial I've encountered this problem when implementing some derived state caching (based on both context & props) in a real world application. It caught me completely off guard because I was relaying on the fact that `React.memo` should always provide me same props (referentially equal) and I've only checked against that - skipping shallowEqual because I was trusting React to perform this one on my behalf.

Reply:
Can you change the test to rely on `props.something` rather than `props` itself? Relying on referential identity of `props` seems a bit like assuming too much to me

Reply:
Comparing against props referential identity isnt that unusual - https://github.com/reduxjs/react-redux/blob/fac9ad19d17000f5b9fdd52fc9843225b807845b/src/components/connectAdvanced.js#L142 .

I can't change this test to rely on `props.something` because this issue is specifically about the fact that props are not referentially equal in this case (sorry if I havent made that clear before). Anything on the props will be referentially equal as it pass `shallowEqual` test.

My point is that the current behaviour is **inconsistent** here and IMHO that's not a good thing because it makes the mental model about this a little bit unreliable.

Have you checked out my codesandbox which compares `React.memo(Component)` & `React.memo(Component, shallowEqual)`? I strongly believe that they should behave exactly the same and should not be distinguishable from user's perspective, as React.memo implementation conceptually looks like:
```
React.memo = (Component, compare = shallowEqual) => {
 // ...
}
```
Therefore supplying the second argument implicitly (relaying on default params) and explicitly should result in the very same observable behaviour.

When I've encountered this issue in my code I've supplied shallowEqual as 2nd argument by hand, just so I could put a debugger statement quicker in it's implementation to see why my component is rerendering and to my surprise it has fixed "the issue" which was at very least really surprising (knowing that `React.memo` uses `shallowEqual` by default)